content/blog/2019-12-02-equivalence-modulo-inputs/index.md

```diff
@@ -0,0 +1,400 @@
+which, since we're only checking equivalence over a particular set of
+inputs, gives an efficient procedure for (2).
+
+[rice]: https://en.wikipedia.org/wiki/Rice%27s_theorem
+
+
+## EMI in Practice: Orion
+
+Le & al. realize the promise of Equivalence Modulo Inputs by implementing Orion,
+a bug-finding tool for C compilers.
+Given a seed program and an input set, it generates EMI-variants of the seed
+program and then checks if a compiler configuration is EMI-valid with respect
+to these.
+
+To generate EMI-variants, Orion uses code coverage information provided
+by tools such as `gcov` to modify unexecuted parts of the seed program.
+Intuitively, this procedure generates EMI-variants of the seed program
+since unexecuted statements should not affect the output of the compiled program.
+
+Specifically, Orion probabilistically deletes unexecuted statements of
+seed programs to generate EMI-variants.
+The authors consider other mutation strategies as part of future work,
+but as we'll see in the evaluation section below, the simple strategy of
+deleting statements works well in practice.
+
+Orion's EMI-variant generation algorithm is sketched below in `gen_variant`.
+
+```Python
+def prune_visit(prog, statement, coverage_set):
+  # probabilistically delete unexecuted statement
+  if statement not in coverage_set and flip_coin(statement):
+    prog.delete(statement)
+
+  # otherwise, traverse its children
+  else:
+    for child in statement.children:
+      prune_visit(prog, child, coverage_set)
+
+def gen_variant(prog, coverage_set):
+  emi_variant = clone(prog)
+  for statement in emi_variant:
+    prune_visit(emi_variant, statement, coverage_set)
+
+  return emi_variant
+```
+
+`gen_variant` takes as input a seed program and the coverage set&mdash;the
+set of all statements executed for some input in the input set.
+It clones the program into `emi_variant` and then uses `prune_visit` to
+probabilistically delete unexecuted statements.
+
+With its EMI-variant generation algorithm outlined, we can now sketch
+the algorithm by which Orion validates C compilers.
+
+```Python
+def validate(compiler, prog, input_set):
+  # compile with no optimizations
+  out_prog = compiler.compile(prog, NO_OPTIMIZATION)
+
+  # generate reference output
+  in_out_set = [(i, out_prog.execute(i)) for i in input_set]
+
+  # get coverage info
+  # a statement is considered covered if it was executed
+  # by the program on any input
+  coverage_set = set()
+  for i in input_set:
+    coverage_set = union(coverage_set, coverage(prog, i))
+
+  for i in range(MAX_ITER):
+    emi_variant = gen_variant(prog, coverage_set)
+
+    for config in compiler.configurations:
+      out_emi_variant = compiler.compile(emi_variant, config)
+
+      # check if compiled EMI variant is equivalent over all inputs
+      for i, o in in_out_set:
+        # compiler is not EMI-valid!
+        if out_emi_variant.execute(i) != o:
+          report_bug(compiler, config, prog, emi_variant)
```

Comment:
Maybe this is silly, but it seems like the report would want to include `i`, `o`, and the actual output for the variant?